tests/robustness/traffic/kubernetes.go

```diff
@@ -155,31 +152,56 @@ func (t kubernetesTraffic) generateKey() string {
 	return fmt.Sprintf("/registry/%s/%s/%s", t.resource, t.namespace, stringutil.RandString(5))
 }
 
-func (t kubernetesTraffic) Range(ctx context.Context, c *RecordingClient, key string, withPrefix bool) (*clientv3.GetResponse, error) {
-	ctx, cancel := context.WithTimeout(ctx, RequestTimeout)
-	resp, err := c.Range(ctx, key, withPrefix)
-	cancel()
+type KubernetesRequestType string
+
+const (
+	KubernetesDelete KubernetesRequestType = "delete"
+	KubernetesUpdate KubernetesRequestType = "update"
+	KubernetesCreate KubernetesRequestType = "create"
+)
+
+type kubernetesClient struct {
+	client *RecordingClient
+}
+
+func (k kubernetesClient) List(ctx context.Context, key string) (*clientv3.GetResponse, error) {
+	resp, err := k.client.Range(ctx, key, true)
+	if err != nil {
+		return nil, err
+	}
 	return resp, err
 }
 
-func (t kubernetesTraffic) Create(ctx context.Context, c *RecordingClient, key, value string) error {
-	return t.Update(ctx, c, key, value, 0)
+func (k kubernetesClient) OptimisticDelete(ctx context.Context, key string, expectedRevision int64) (*mvccpb.KeyValue, error) {
+	return k.optimisticOperationOrGet(ctx, key, clientv3.OpDelete(key), expectedRevision)
 }
 
-func (t kubernetesTraffic) Update(ctx context.Context, c *RecordingClient, key, value string, expectedRevision int64) error {
-	ctx, cancel := context.WithTimeout(ctx, RequestTimeout)
-	err := c.CompareRevisionAndPut(ctx, key, value, expectedRevision)
-	cancel()
-	return err
+func (k kubernetesClient) OptimisticUpdate(ctx context.Context, key, value string, expectedRevision int64) (*mvccpb.KeyValue, error) {
+	return k.optimisticOperationOrGet(ctx, key, clientv3.OpPut(key, value), expectedRevision)
 }
 
-func (t kubernetesTraffic) Delete(ctx context.Context, c *RecordingClient, key string, expectedRevision int64) error {
-	ctx, cancel := context.WithTimeout(ctx, RequestTimeout)
-	err := c.CompareRevisionAndDelete(ctx, key, expectedRevision)
-	cancel()
+func (k kubernetesClient) OptimisticCreate(ctx context.Context, key, value string) error {
+	_, err := k.client.Txn(ctx, []clientv3.Cmp{clientv3.Compare(clientv3.ModRevision(key), "=", 0)}, []clientv3.Op{clientv3.OpPut(key, value)}, nil)
 	return err
 }
 
+func (k kubernetesClient) optimisticOperationOrGet(ctx context.Context, key string, operation clientv3.Op, expectedRevision int64) (*mvccpb.KeyValue, error) {
```

Comment:
Please add a comment or link to describe the concept of Kubernetes optimistic operation.

Reply:
Fone.